qiskit/algorithms/minimum_eigen_solvers/adapt_vqe.py

```diff
@@ -0,0 +1,338 @@
+        commutators = []
+        for exc in self._excitation_pool:
+            # The excitations operators are applied later as exp(i*theta*excitation).
+            # For this commutator, we need to explicitly pull in the imaginary phase.
+            commutator = 1j * ((operator @ exc) - (exc @ operator)).reduce()
+            commutators.append(commutator)
+
+        wave_function = self._solver.ansatz.assign_parameters(theta)
+
+        res = eval_observables(
+            self._solver.quantum_instance, wave_function, commutators, expectation
+        )
+        return res
+
+    @staticmethod
+    def _check_cyclicity(indices: List[int]) -> bool:
+        """
+        Auxiliary function to check for cycles in the indices of the selected excitations.
+
+        Args:
+            indices: The list of chosen gradient indices.
+
+        Returns:
+            Whether repeating sequences of indices have been detected.
+        """
+        cycle_regex = re.compile(r"(\b.+ .+\b)( \b\1\b)+")
+        # reg-ex explanation:
+        # 1. (\b.+ .+\b) will match at least two numbers and try to match as many as possible. The
+        #    word boundaries in the beginning and end ensure that now numbers are split into digits.
+        # 2. the match of this part is placed into capture group 1
+        # 3. ( \b\1\b)+ will match a space followed by the contents of capture group 1 (again
+        #    delimited by word boundaries to avoid separation into digits).
+        # -> this results in any sequence of at least two numbers being detected
+        match = cycle_regex.search(" ".join(map(str, indices)))
+        logger.debug("Cycle detected: %s", match)
+        logger.info("Alternating sequence found. Finishing.")
+        # Additionally we also need to check whether the last two numbers are identical, because the
+        # reg-ex above will only find cycles of at least two consecutive numbers.
+        # It is sufficient to assert that the last two numbers are different due to the iterative
+        # nature of the algorithm.
+        return match is not None or (len(indices) > 1 and indices[-2] == indices[-1])
+
+    @staticmethod
+    def _log_gradients(
+        iteration: int,
+        cur_grads: Tuple[float, PauliSumOp],
+        max_grad: Tuple[float, PauliSumOp | None],
+    ):
+        if logger.isEnabledFor(logging.INFO):
+            gradlog = []
+            gradlog.append(
+                f"\nGradients in iteration #{str(iteration)}\n"
+                "ID: Excitation Operator: Gradient  <(*) maximum>"
+            )
+            for i, grad in enumerate(cur_grads):
+                gradlog.append(f"\n{str(i)}: {str(grad[1])}: {str(grad[0])}")
+                if grad[1] == max_grad[1]:
+                    gradlog.append("\t(*)")
+            logger.info(",".join(gradlog))
+
+    def compute_minimum_eigenvalue(
+        self,
+        operator: OperatorBase,
+        aux_operators: Optional[ListOrDict[PauliSumOp]] = None,
```

Comment:
Is there a reason this only supports `PauliSumOp`s as operators? The minimum eigensolvers interface accepts any `OperatorBase` usually 🙂